Implement GraphQL integration for fetching plan section questions and…

fraserclark opened the pull request:
## Description

#366 
This implements GraphQL data fetching for the section questions page.

**Key Changes:**
- Added GraphQL queries for `PlanSectionQuestions`, `Section`, and `Plan` data
- Replaced hardcoded question data with dynamic fetching
- Updated breadcrumb navigation to use real project/plan data
- Added tests

**Questions:**

How do I orderBy display order? I noticed none of the other graphql queries are ordering?

Another thing I noticed was when I enter random numbers into the DMP part 

my actual url 
/en-US/projects/6/dmp/6
to
/en-US/projects/6/dmp/888/s/6

It still works - my graphql query just looks at the section id - does it need to be more sophisticated or will middleware stop this?

Loading States- The current loading state is basic ("Loading questions..."). Do we have a more sophisticated loading skeleton or should we throw in the backlog?


## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?
Added to the test suite and in the browser 

## Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [X] I updated the CHANGELOG.md and added documentation if necessary
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have completed manual or automated accessibility testing for my changes
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

fraserclark commented on graphql/queries/questions.query.graphql:
```diff
@@ -4,6 +4,20 @@ query QuestionsDisplayOrder($sectionId: Int!) {
   }
 }
 
+query PlanSectionQuestions($sectionId: Int!) {
```
I assume some thing like this? based on what I read?

```
query PlanSectionQuestions($sectionId: Int!) {
  questions(sectionId: $sectionId, orderBy: { displayOrder: ASC }) {
    id
    questionText
    displayOrder
    guidanceText
    requirementText
    sampleText
    sectionId
    templateId
    isDirty
  }
}
```

briri replied:
Not every query supports pagination/sorting at the moment. For reference, you can always check the [schema files in the backend](https://github.com/CDLUC3/dmsp_backend_prototype/blob/8cf05c431cf56ec05e848e1a89c21f119ad6899e/src/schemas/question.ts#L6) (make sure you're on the development branch). The ones that support pagination and sorting will have `paginationOptions` as a possible input argument. E.g.
```
    "Get all of the user's projects"
    myProjects(term: String, paginationOptions: PaginationOptions): ProjectSearchResults
```

The backend currently doesn't specify any sort order for this particular query (it should probably sort by displayOrder though). I added [a ticket](https://github.com/CDLUC3/dmsp_backend_prototype/issues/324) to do so.

fraserclark commented on app/[locale]/projects/[projectId]/dmp/[dmpid]/s/[sid]/page.tsx:
```diff
@@ -20,52 +24,89 @@ interface Question {
+    isAnswered: false
+  })) || [];
+
+  if (questionsLoading || sectionLoading || planLoading) {
```
Probably need better handling?

jupiter007 commented:
@fraserclark the left sidebar is missing from the page. I believe it was supposed to be a part of this ticket when you look at the wireframe: https://www.figma.com/proto/RRtoRLyr5Spf0KiyRfEclt/DMP-Tool-Wireframes?node-id=1575-895&t=pGJl655GG7A2dvdJ-0&scaling=scale-down-width&page-id=177%3A504&starting-point-node-id=177%3A626&show-proto-sidebar=1&content-scaling=fixed

Other than that, I only had a few comments. Looks like the requirements are still hard-coded because we don't have them available yet?

jupiter007 commented:
Also, in response to your question
```
my actual url
/en-US/projects/6/dmp/6
to
/en-US/projects/6/dmp/888/s/6

It still works - my graphql query just looks at the section id - does it need to be more sophisticated or will middleware stop this?
```
about how to handle non-existent `sectionId`, I think you should check whether you have an empty `questionsData` set and redirect back to the previous page with a message. Middleware will not check for a valid `sectionId` in the path. That should be done in the page component itself I think.

fraserclark commented:
I put some logic in to validate section belongs to plan and 404 if it does not.

I added a v1 of the sidebar - think will need to revisit it.

Also fixed a few links that I noticed had broken on the project overview page

jupiter007 commented:
@fraserclark, I saw that you requested another review of this ticket, but I think you might've missed this comment of mine:
- In `app/[locale]/projects/[projectId]/dmp/[dmpid]/s/[sid]/page.tsx`, `notFound()` is a server-side function. Instead, you could display an error message and redirect user to the previous page. Something like "Plan not found" or something.
- Also on the same page, it looks like `template_name` is not being used.

jupiter007 commented:
> @fraserclark, I saw that you requested another review of this ticket, but I think you might've missed this comment of mine:
> 
> * In `app/[locale]/projects/[projectId]/dmp/[dmpid]/s/[sid]/page.tsx`, `notFound()` is a server-side function. Instead, you could display an error message and redirect user to the previous page. Something like "Plan not found" or something.
> * Also on the same page, it looks like `template_name` is not being used.

Oh I just checked. I guess I was mistaken about `notFound()` function. I thought it was server-only, but I read you can use it on the client side.

- If you can just address the issue with `template_name` and `navBottom` that don't appear to be used in the component.

jupiter007 commented:
Also, thanks for adding the left sidebar sections @fraserclark

andrewebdev commented on app/[locale]/projects/[projectId]/dmp/[dmpid]/s/[sid]/page.tsx:
```diff
@@ -74,76 +152,124 @@ const PlanOverviewSectionPage: React.FC = () => {
+                            className={styles.progressIndicator}
+                            aria-label={`Question status: ${question.isAnswered ? 'Completed' : 'Not started'}`}
+                          >
+                            <svg
```
When adding icons, we need to add them to the `DmpIcon` component that we already have. Adding the SVG directly in the page isn't very re-usable, and cumbersome to maintain.

andrewebdev commented on app/[locale]/projects/[projectId]/dmp/[dmpid]/s/[sid]/page.tsx:
```diff
@@ -156,48 +282,48 @@ const PlanOverviewSectionPage: React.FC = () => {
+              className={styles.bestPracticesLinks}>
               <Link href="/best-practices/sharing">
                 Data sharing
                 <svg width="20" height="20" viewBox="0 0 20 20"
```
Should change this to use `DmpIcon`

andrewebdev commented on app/[locale]/projects/[projectId]/dmp/[dmpid]/s/[sid]/page.tsx:
```diff
@@ -156,48 +282,48 @@ const PlanOverviewSectionPage: React.FC = () => {
 
               <Link href="/best-practices/preservation">
                 Data preservation
                 <svg width="20" height="20" viewBox="0 0 20 20"
```
Change to `DmpIcon`

andrewebdev commented on app/[locale]/projects/[projectId]/dmp/[dmpid]/s/[sid]/page.tsx:
```diff
@@ -156,48 +282,48 @@ const PlanOverviewSectionPage: React.FC = () => {
 
               <Link href="/best-practices/protection">
                 Data protection
                 <svg width="20" height="20" viewBox="0 0 20 20"
```
`DmpIcon` (apologies for repetition, just adding them so we don't forget

andrewebdev commented on app/[locale]/projects/[projectId]/dmp/[dmpid]/s/[sid]/page.tsx:
```diff
@@ -156,48 +282,48 @@ const PlanOverviewSectionPage: React.FC = () => {
 
               <Link href="/best-practices/all">
                 All topics
                 <svg width="20" height="20" viewBox="0 0 20 20"
```
`DmpIcon`

andrewebdev commented on app/[locale]/projects/[projectId]/page.tsx:
```diff
@@ -314,7 +322,10 @@ const ProjectOverviewPage: React.FC = () => {
+                        href={routePath('projects.dmp.show', {
+                          projectId: String(projectId),
+                          dmpId: planId
+                        })}
```
I think @jupiter007 also made this fix in a PR recently. Make sure to re-merge dev and resolve conflicts into your branch just to be sure.

andrewebdev reviewed (changes requested):
Just a couple of comments from my side. We have to start using the `DmpIcon` rather than copy/paste svg directly into pages. It makes it hard to maintain the code, or manage changes to icons in future.

jupiter007 commented on app/[locale]/projects/[projectId]/page.tsx:
```diff
@@ -261,8 +262,8 @@ const ProjectOverviewPage: React.FC = () => {
             {project.plans.map((plan) => {
-              // extract dmp id from the full dmpId
-              const doiId = plan?.dmpId?.match(/(?:https?:\/\/doi\.org\/)(.+)/)?.[1] ?? '';
+              // Use the plan ID for routing, not the DOI extraction
```
L99: Just an FYI. I built the `Question Details` page already but moved it under the `Section` id, so it's at `projects/[projectId]/dmp/[dmpid]/s/[sid]/q/[qid]`. The `routePath` for that is `projects.dmp.question.detail`
